_posts/2019/2019-07-16-why-we-stopped-stoplight.md

```diff
@@ -0,0 +1,79 @@
+This way users can learn about our API and try it out in one place instead of having to switch between tools.
+In short, we wanted to make life easier for developers 😉.
+
+## Goodbye Asciidoctor - Hello OpenAPI
+
+To generate our static [HTML documentation](http://docs.beyondshop.cloud/){:target="_blank"} we use [Spring REST Docs](https://spring.io/projects/spring-restdocs){:target="_blank"}, a test-driven approach to produce documentation for RESTful services.
+For Stoplight to be our new tool, our backend team had to make some considerable effort as the tool is based on [OpenAPI](https://swagger.io/docs/specification/about/){:target="_blank"}.
+For example, they built an [API specification support to Spring REST Docs](https://github.com/ePages-de/restdocs-api-spec){:target="_blank"} that, amongst others, supports OpenAPI.
+
+And what's more, our developers had to:
+
+* make adjustments to our microservices to meet the requirements of OpenAPI
+* build a deployment pipeline
+* adjust the API tests
+* support our Technical Communicators with the fine tuning.
+
+(To be honest, it was not really a goodbye to Asciidoctor.
+As it is an integral part to publish our internal developer documentation, it had to remain.)
+
+## Soft launch
+
+Automatically turning an OpenAPI specification into an API Reference, making the navigation a lot easier, introducing interactivity by sending test requests during document usage, a code generator for a multitude of languages and frameworks, a search... all that made our eyes shine!
+And a bunch of Frontend developers were happily using the documentation from the very beginning when we internally launched it.
+That was a good opportunity to try and test the docs before we let any external developer on it. 
+
+## Practice makes perfect
+
+A good six months had passed with a lot of shaping still to be done here and there.
+When we thought, we were ready for public launch, we invested some time and checked the API reference of the static HTML document with what we had in Stoplight.
+From the time we started with the tool until then, the API reference had changed a lot.
+New endpoints had been crafted, and the API had grown massively.
+No wonder that we were wondering why, during our final check, lots of these endpoints were missing in Stoplight.
+
+We asked our developers to investigate and the result somewhat surprised us:
+OpenAPI V2 cannot provide different schemas and/or payload examples for a single URI.
+Wow, that was some news!
+So much effort, and now this!
+As our API handles many endpoints with a single URI such as products and variation products, there was no quick and easy solution to solve this issue.
+And releasing an incomplete documentation was no option for us.
+
+## What now?
+
+We figured that OpenAPI V3 caters for different schemas/examples for one URI.
+But implementing that requires backend resources for the rework.
```

Comment:
Again, we could generate it with minimal effort thanks to the open-source project, where contributors added most of the OpenAPI 3 support.

Reply:
Yes, true. But what was communicated was that it still requires dev resources to that and Timo doesn't want to invest them, right?

Reply:
Yes, true. But what was communicated was that it still requires dev resources to that and Timo doesn't want to invest them, right?

Reply:
Similar to throwing out OpenAPI 2.
Generating OpenAPI 3 is basically some few more lines of config per service.

Timo didn't want to invest, because we already identified some shortcomings of the OpenAPI 3 implementation (remember: it was contributed in our open-source project by externals, it works for the most part/their use-cases, but might be incomplete) and whether we will eventually get a benefit from importing it in stoplight.